Adds ROSIndex build configuration

hidmic opened the pull request:
Connected to ros-infrastructure/ros_buildfarm#568. This pull request makes use of the job support provided by [ros-infrastructure/ros_buildfarm#576](https://github.com/ros-infrastructure/ros_buildfarm/pull/576) and the containerization improvements in [ros2/rosindex#83](https://github.com/ros2/rosindex/pull/83) to setup daily rosindex builds.

A note of warning: upload repository ros-infrastructure/index.ros.org has not been configured yet with buildfarm SSH keys as deploy keys with write access.

hidmic committed:
Adds ROSIndex build configuration

hidmic commented:
@nuclearsandwich yet another PR for you to take a look at :)

nuclearsandwich commented on doc-independent-rosindex.yaml:
```diff
@@ -0,0 +1,20 @@
+documentation_type: external_site
+doc_repositories:
+- https://github.com/ros2/rosindex.git
+jenkins_job_priority: 90
```
we'll want to double check these priority and timeout parameters before merging.

hidmic replied:
Priority for sure, timeout was set after @tfoote [reported 90 minutes to build the whole site](https://github.com/ros2/rosindex/pull/54#issuecomment-424290434).I took a 30 minutes buffer, just in case.

nuclearsandwich commented on doc-independent-rosindex.yaml:
```diff
@@ -0,0 +1,20 @@
+- https://github.com/ros2/rosindex.git
+jenkins_job_priority: 90
+jenkins_job_timeout: 120
+targets:
```
Since this job runs mostly in a container from the repository source we should either remove the target stanza, or if that's not easily doable comment that it is ignored for this documentation type.

hidmic replied:
I guess we could change the `BuildFile` base class definition to make that stanza optional on a case by case basis. I wonder though, do `make_target` doc_independent jobs make use of it?

nuclearsandwich commented on index.yaml:
```diff
@@ -37,6 +37,7 @@ distributions:
       default: kinetic/source-build.yaml
 doc_builds:
   independent-packages: doc-independent-build.yaml
+  independent-rosindex: doc-independent-rosindex.yaml
```
We can probably drop `independent` from the name and just use rosindex which will produce the `doc_rosindex` job on the buildfarm.

hidmic replied:
Ok, makes sense.

nuclearsandwich commented:
I'm not actually sure whether we need or want a rosindex config in the example branch (which `master` is for this repository).

We will want some version of this change contributed against the `production` branch (which has a divergent history FYI @hidmic)

hidmic committed:
Changes ROSIndex job configuration file name and key.

hidmic commented:
OH, I didn't know you had a different branch convention. Let me fix it.

hidmic committed:
Changes rosindex doc type to docker_build.

nuclearsandwich committed:
Update credential ID for production deployment.

nuclearsandwich committed:
Remove unused upload config

The commented out config values are for make_target jobs and not applicable here.

dirk-thomas commented on doc-rosindex.yaml:
```diff
@@ -0,0 +1,16 @@
+jenkins_job_timeout: 120
+targets:
+  ubuntu:
+    trusty:
```
Is there a reason to run this job on Trusty? If not I would prefer it to run on a more up-to-date distro to avoid it using a platform which will be EOL in less then half a year.

hidmic replied:
Hmm, this is something that @nuclearsandwich already brought up [here](https://github.com/ros-infrastructure/ros_buildfarm_config/pull/123#discussion_r226365373). Since the Docker container is source side provided, we do not really control the target platform. We could either make it configurable through Dockerfile argument (but IMHO that's a hard constrain on the user) or remove the stanza, changing the configuration classes inside ros_buildfarm to support making it optional (it's not, see [here](https://github.com/ros-infrastructure/ros_buildfarm/blob/master/ros_buildfarm/config/build_file.py#L57)).

dirk-thomas replied:
If this part isn't being used than the build file shouldn't need to provide it.

hidmic replied:
Ok, optional it is. I'll update ros-infrastructure/ros_buildfarm#576.

tfoote commented:
http://build.ros.org/job/doc_rosindex/5/console

The build of the index took 60% of the RAM on an 8GB machines

This caused 2 consecutive build agents under load to fall over due to OOM errors.

I monitored memory during the build and have notes here:
https://gist.github.com/tfoote/9af6d68ebe4dc077691c2d03718617e9

I'm not sure what we can do about this, for now we may need to run things on an isolated machine.

dirk-thomas commented:
The commit the build generates is also extremely huge. I would suggest to:
* ~~purge the [automatic commit](https://github.com/ros-infrastructure/index.ros.org/commit/a0d1f4ac294d7985b40c1b2a824b53217d6c1e98) from the `gh-pages` branch~~ looking at individual diff it seems the page template was updated and each page needed to be touched and
* update the generation process ~~strip~~  ignore the non-reproducible part of the artifacts (likely timestamp etc.) before committing the changes: so if the only diff is the snippet `generated on YYY-MM-DD` then the file should not be committed / the change be reverted. Then the modified timestamp on the website will actually reflect when "something" has changed last on that page.

hidmic commented:
> The build of the index took 60% of the RAM on an 8GB machines

I took a peek inside Jekyll. It starts writing to disk only once it has rendered the whole site in-memory (see [here](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/site.rb#L68)). But I'm curious about your notes @tfoote, does it really start eating more memory when it logs rendering stats??? 

Anyways, a few things I think we can do about this (besides throwing more silicon at it):

- Remove all `--trace` and `--verbose` flags from `jekyll build` call sites. It's easy enough, though I'd be quite surprised if that actually improves anything.

- Monkey patch Jekyll a bit (using a mixin). Both [rendering](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/site.rb#L479) and [writing](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/site.rb#L209) stages work with [Document](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/document.rb#L4) instances. And I presume rendered content explains a large portion of the process memory footprint. If I'm right, then we just need to start swapping it to disk earlier. So what if we make the [output accessor](https://github.com/jekyll/jekyll/blob/master/lib/jekyll/document.rb#L9) write to a temporary file in disk? We can keep around weakrefs to its content in memory to speed up accesses and optimize the write method to just hardlink the file at the destination path. 

> ...update the generation process strip ignore the non-reproducible part of the artifacts (likely timestamp etc.) before committing the changes...

Do we even need that timestamp? Why not just remove it or make it optional? It'd definitely make stripping easier.

chapulina commented on doc-rosindex.yaml:
```diff
@@ -0,0 +1,16 @@
+---
+documentation_type: docker_build
+doc_repositories:
+- https://github.com/ros2/rosindex.git
```
```suggestion
- https://github.com/ros-infrastructure/rosindex.git
```

nuclearsandwich replied:
The repo hadn't been moved when I dropped the link. :+1:

nuclearsandwich reviewed (approved):
There are no problems with the build configuration. We'll need to address Jekyll's memory consumption in order for builds to be reliable under load.

chapulina committed:
Updates upload repository URL

Co-Authored-By: hidmic <hid.michel@gmail.com>

hidmic commented:
Awesome, thanks @dirk-thomas @nuclearsandwich and @chapulina for taking a look!
